### PR TITLE
acl on Provenance data queries

### PR DESCRIPTION
**Pull request by cendhu**

### CODE REVIEW

File: internal/bcdb/provenance_query_processor.go [hunk 1]
```diff
@@ -209,15 +241,15 @@ func (p *provenanceQueryProcessor) aclCheckForUserOperation(querierUserID, targe
...
 	isAdmin, err := p.identityQuerier.HasAdministrationPrivilege(querierUserID)
 	if err != nil {
 		return err
 	}
 
-	if !isAdmin && querierUserID != targetUserID {
+	if !isAdmin {
 		return &errors.PermissionErr{
-			ErrMsg: "The querier is neither an admin nor requesting reads performed by the querier. Only an admin can query operations performed by other users.",
+			ErrMsg: "The querier [" + querierUserID + "] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor.go:252]: make sure that handlers that consume this error return 403

File: internal/bcdb/provenance_query_processor_test.go [hunk 2]
```diff
@@ -489,16 +515,30 @@ func TestGetDeletedValues(t *testing.T) {
...
 				Values: nil,
 			},
 		},
+		{
+			name:             "fetching all deleted values key1 with non-admin user will result in an error",
+			dbName:           "db1",
+			key:              "key1",
+			user:             "user1",
+			expectedEnvelope: nil,
+			expectedError:    "The querier [user1] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:529]: match the error type as well

File: internal/bcdb/provenance_query_processor_test.go [hunk 3]
```diff
@@ -552,18 +598,36 @@ func TestGetPreviousValues(t *testing.T) {
...
+			name:   "fetching the previous value of key1 at version{Blk 3, txNum 0} by an non-admin user will result in an error",
+			dbName: "db1",
+			key:    "key1",
+			user:   "user1",
+			version: &types.Version{
+				BlockNum: 3,
+				TxNum:    0,
+			},
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:618]: match the error type as well, etc...

File: internal/bcdb/provenance_query_processor_test.go [hunk 4]
```diff
@@ -625,12 +696,29 @@ func TestGetNextValues(t *testing.T) {
...
+			name:   "fetching next value of key1 at version {Blk 2, txNum 0} by an non-admin user will result in an error",
+			dbName: "db1",
+			key:    "key1",
+			user:   "user1",
+			version: &types.Version{
+				BlockNum: 2,
+				TxNum:    0,
+			},
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:709]: as above, match error type as well

File: internal/bcdb/provenance_query_processor_test.go [hunk 5]
```diff
@@ -683,12 +778,29 @@ func TestGetValueAt(t *testing.T) {
...
+			name:   "fetching value of key1 at a particular version by an non-admin user will result in an error",
+			dbName: "db1",
+			key:    "key1",
+			user:   "user1",
+			version: &types.Version{
+				BlockNum: 3,
+				TxNum:    0,
+			},
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:791]: as above, match error type as well

File: internal/bcdb/provenance_query_processor_test.go [hunk 6]
```diff
@@ -741,12 +860,29 @@ func TestGetMostRecentValueAtOrBelow(t *testing.T) {
...
+			dbName: "db1",
+			key:    "key1",
+			user:   "user1",
+			version: &types.Version{
+				BlockNum: 2,
+				TxNum:    5,
+			},
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
+		},
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:874]: as above, match error type as well

File: internal/bcdb/provenance_query_processor_test.go [hunk 7]
```diff
@@ -779,16 +921,30 @@ func TestGetReaders(t *testing.T) {
...
 			},
 		},
+		{
+			name:            "fetching readers of key1 by an non-admin user will result in an error",
+			dbName:          "db1",
+			key:             "key1",
+			user:            "user1",
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
+		},
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:936]: as above, match error type as well

File: internal/bcdb/provenance_query_processor_test.go [hunk 8]
```diff
@@ -821,14 +983,29 @@ func TestGetWriters(t *testing.T) {
...
 				WrittenBy: nil,
 			},
 		},
+		{
+			name:            "fetching readers of key1 by an non-admin user will result in an error",
+			dbName:          "db1",
+			key:             "key1",
+			user:            "user1",
+			expectedPayload: nil,
+			expectedError:   "The querier [user1] is not an admin. Only an admin can query historical data",
```
Comment (tock-ibm) [inline internal/bcdb/provenance_query_processor_test.go:997]: as above, match error type as well

**Review by tock-ibm — COMMENTED** [review]
@cendhu Add an issue to improve provenance integration tests to check ACLs please

File: internal/httphandler/provenance_request_handler.go [hunk 9]
```diff
@@ -117,7 +117,7 @@ func (p *provenanceRequestHandler) getDataReaders(w http.ResponseWriter, r *http
 	}
 	query := payload.(*types.GetDataReadersQuery)
 
-	response, err := p.db.GetReaders(query.DbName, query.Key)
+	response, err := p.db.GetReaders(query.UserId, query.DbName, query.Key)
 	if err != nil {
 		processInternalError(w, r, err)
 		return
```
[the remark this answers is not in the document]
Comment (cendhu) [reply internal/httphandler/provenance_request_handler.go]: Done

File: internal/bcdb/provenance_query_processor.go (hunk 1 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 2 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 3 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 4 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 5 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 6 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

File: internal/bcdb/provenance_query_processor_test.go (hunk 7 above)
Comment (cendhu) [reply internal/bcdb/provenance_query_processor_test.go]: Done

**Review by tock-ibm — APPROVED** [review]

**Pull request merged**